### PR TITLE
Support remote port forwarding

### PR DESCRIPTION
**Pull request by jonah-iden**
#### What it does

Builds on top #13372. So that PR should probably be merged first

This adds the ability to the remote feature to forward ports to the remote system. Works for both ssh and dev-container connections.
Adds a new Port Widget to manage forwarded Ports.
grafik

#### How to test

1. Connect theia to a remote system either through ssh or the dev-container support
2. start some application on the remote system listening to a port. Like a simple node.js server for example
3. go to views->open-view...->Ports to open the Port forwarding UI
4. forward the port the application is listening to
5. connection to the application should work

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by jonah-iden** [comment]
@JonasHelming @planger here is the dev-container follow up PR for the remote port forwarding at runtime. Should work in general for the remote feature. So both fort SSH and Dev-Containers.

**Comment by planger** [comment]
Cool, thanks! I'm happy to take a look at it next week.

**Review by planger — CHANGES_REQUESTED** [review]
Thank you for adding this great new feature to the remote container support! Overall it looks and works very well.

I'd like to mention a few observations below and in the code inline.

The padding of the message "No forwarded ports..." in the widget seems to be missing. E.g. in the Problem view, when no problems are found, this message has a padding of 13px, whereas in the Ports view, this message has no padding at all.

There seems to be a minor focus issue in the Port view:
```
root WARN Widget was activated, but did not accept focus after 2000ms: port-forwarding-widget
```

When specifying `localhost:<port>` in the port column, it correctly binds only to `localhost`, thus it is not reachable with the external IP address. If specified without e.g. `localhost:`, it is reachable from the outside. This is great! :+1:
However in the UI, this difference doesn't show. Afaik, it also doesn't show in VS Code, but imho it should (e.g. to avoid wondering why you can't access the port when it is bound to a specific IP address). This isn't the most important thing though and certainly shouldn't block merging this. Just thought it might be easy to add in some way.

File: packages/dev-container/src/electron-node/docker-container-creation-service.ts
Comment (jonah-iden) [reply packages/dev-container/src/electron-node/docker-container-creation-service.ts:39]: created an issue for it https://github.com/eclipse-theia/theia/issues/13503

File: packages/dev-container/src/electron-node/docker-container-creation-service.ts
Comment (jonah-iden) [reply packages/dev-container/src/electron-node/docker-container-creation-service.ts:1]: ups i guess i made a mistake while rebasing. I'll delete this one. Thanks for finding

**Comment by JonasHelming** [comment]
@jonah-iden Is this ready for a re-review?

**Comment by jonah-iden** [comment]
@JonasHelming im so sorry. Yes it is. I totally forgot to request the rereview

**Review by planger — COMMENTED** [review]
Thanks @jonah-iden for the improvements!

These two observations of my [previous review](https://github.com/eclipse-theia/theia/pull/13439#pullrequestreview-1945578958), however, can still be observed and should hopefully be easy to address, before we can merge this change:

> The padding of the message "No forwarded ports..." in the widget seems to be missing. E.g. in the Problem view, when no problems are found, this message has a padding of 13px, whereas in the Ports view, this message has no padding at all.
> 
> There seems to be a minor focus issue in the Port view:
> ```
> root WARN Widget was activated, but did not accept focus after 2000ms: port-> forwarding-widget
> ```

My earlier comment of indicating that a port is bound to an IP address can be postponed, I guess, as it is also not shown in VS Code.

**Comment by jonah-iden** [comment]
ups sorry. I guess i must have overlooked those comments. I'll take another look later today

**Comment by planger** [comment]
Great, thank you very much @jonah-iden!

**Comment by jonah-iden** [comment]
@planger i changed the requested stuff. Also when not giving an address like localhost, it should now show as 0.0.0.0 to indicate its reachable from the outside. Is that ok like that?

**Pull request merged**
